src/Traits/HasPermissions.php

```diff
@@ -18,18 +18,19 @@
      */
     public function givePermissionTo(...$permissions)
     {
-        $permissions = collect($permissions)
+        collect($permissions)
             ->flatten()
             ->map(function ($permission) {
                 return $this->getStoredPermission($permission);
             })
             ->each(function ($permission) {
                 $this->ensureModelSharesGuard($permission);
+                if (! $this->hasPermissionTo($permission->name)) {
```

Comment:
Maybe we can use a `filter` method on the collection here to filter out permissions that are already given to the user? Same for the `assignRole` method.